File: 백준/Gold/1753. 최단경로/최단경로.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {
    static int V, E;
    static ArrayList<Node>[] graph;
    private static final int INF = Integer.MAX_VALUE;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        V = Integer.parseInt(st.nextToken());
        E = Integer.parseInt(st.nextToken());
        graph = new ArrayList[V + 1];
        for (int i = 1; i <= V; i++) {
            graph[i] = new ArrayList<>();
        }

        int start = Integer.parseInt(br.readLine());
        for (int i = 0; i < E; i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            int w = Integer.parseInt(st.nextToken());
            graph[u].add(new Node(v, w));
        }

        dijkstra(start);
        System.out.println();
    }

    private static void dijkstra(int start) {
        boolean[] visit = new boolean[V + 1];
        int[] dist = new int[V + 1];
        Arrays.fill(dist, INF);
        dist[start] = 0;

        PriorityQueue<Node> pq = new PriorityQueue<>();
        pq.offer(new Node(start, 0));
        while (!pq.isEmpty()) {
            int curIndex = pq.poll().index;
            if (visit[curIndex]) continue;
            visit[curIndex] = true;

            for (Node next : graph[curIndex]) {
                if (dist[next.index] > dist[curIndex] + next.cost) {
                    dist[next.index] = dist[curIndex] + next.cost;
                    pq.offer(new Node(next.index, dist[next.index]));
                }
            }
        }
        for (int i = 1; i < dist.length; i++) {
            System.out.println(dist[i] == INF ? "INF" : dist[i]);
        }
    }

    static class Node implements Comparable<Node> {
        int index, cost;

        Node(int index, int cost) {
            this.index = index;
            this.cost = cost;
        }

        @Override
        public int compareTo(Node o) {
            return Integer.compare(cost, o.cost);
        }
    }
}

File: 백준/Gold/1753. 최단경로/최단경로.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    data class Node(val next: Int, val cost: Int) : Comparable<Node> {
        override fun compareTo(other: Node) = cost.compareTo(other.cost)
    }

    var st = StringTokenizer(readLine())
    val sb = StringBuilder()
    val V = st.nextToken().toInt()
    val E = st.nextToken().toInt()
    val list = Array<ArrayList<Node>>(V + 1) { arrayListOf() }
    val start = readLine().toInt()
    val dist = IntArray(V + 1) { Int.MAX_VALUE }
    for (i in 0 until E) {
        st = StringTokenizer(readLine())
        val u = st.nextToken().toInt()
        val v = st.nextToken().toInt()
        val w = st.nextToken().toInt()
        list[u].add(Node(v, w))
    }

    fun dijkstra(start: Int) {
        val pq = PriorityQueue<Node>()
        val visit = BooleanArray(V + 1)
        dist[start] = 0
        pq.offer(Node(start, dist[start]))

        while (!pq.isEmpty()) {
            val cur = pq.poll()
            if (visit[cur.next]) continue
            visit[cur.next] = true

            for (k in list[cur.next].indices) {
                val n = list[cur.next][k]
                if (!visit[n.next] && dist[n.next] > dist[cur.next] + n.cost) {
                    dist[n.next] = dist[cur.next] + n.cost
                    pq.offer(Node(n.next, dist[n.next]))
                }
            }
        }
    }

    dijkstra(start)

    for (i in 1 until dist.size) {
        sb.append(dist[i].takeIf { it != Int.MAX_VALUE } ?: "INF").append("\n")
    }
    println(sb)
}